### PR TITLE
[16.0][IMP] tracking_manager: allow tracking readonly fields

### PR DESCRIPTION
**Pull request by CRogos**
@AnizR, @Kev-Roche, @sebastienbeau 

What is the reason for the limitation that read_only fields cannot be tracked? 
When there have a look at the Odoo native tracking, there are also read_only fields with tracking enabled:
image

In our case we what to track changes on an enterprise table account.asset, where most of the fields are read_only and can only be changed by a separate wizard. But through this wizard, user changes are possible and therefore tracking might be necessary.

Are there any doubt to remove the read_only filter?

### CODE REVIEW

**Review by legalsylvain — APPROVED** [review]

File: tracking_manager/models/ir_model.py [hunk 1]
```diff
@@ -159,7 +158,6 @@ class IrModelFields(models.Model):
     )
     trackable = fields.Boolean(
         compute="_compute_trackable",
-        store=True,
```
Comment (AnizR) [inline tracking_manager/models/ir_model.py]: Why did you remove store=True?

I think that you may need to add something in the depends of your compute if the field isn't (re-)computed correctly.

Comment (CRogos) [reply tracking_manager/models/ir_model.py]: The trackable attribute is already computed, but with this PR we change the implementation how it is computed. 
Changing the implementation does not change the stored value automatically.
I agree another options would be an migration update script to recalculate all values. 

In think this is the only place where trackable is used. I am not sure if this syntax requires a stored field. But this could also be replaced by an lambda. Beside not changing it, I do not see the benefit of storing this simple value. 

https://github.com/OCA/server-tools/blob/485de9f8f8bba0df1fb905a47bc99e5cccb434a3/tracking_manager/models/ir_model.py#L128-L135

Comment (AnizR) [reply tracking_manager/models/ir_model.py]: I think that a migration script might be needed like you suggested but I am not 100% sure about that..

Comment (CRogos) [reply tracking_manager/models/ir_model.py]: Maybe there is a 3rd opinion over the weekend. I am personally not so a big fan of store=True on computed values. But indeed, changing it has also its risks.

Comment (Kev-Roche) [reply tracking_manager/models/ir_model.py]: I ll prefer to keep it stored because in some cases, we want to save a custom choice of fields tracked, and this independently to its kind (read only, related, ...).

**Review by Kev-Roche — CHANGES_REQUESTED** [review]
I m not sure about this. 
In a way, the behavior of readonly field is known and not changed by users. In order to avoid some noise with too many tracked fields by default, I will prefer to not include the readonly fields as tracked. 
IMHO, maybe just set as tracked the relevant fields to follow, as you mention in your specific case, will be enough.

File: tracking_manager/models/ir_model.py (hunk 1 above)
Comment (CRogos) [reply tracking_manager/models/ir_model.py]: I think "trackable" is only used as a filter when showing all trackable fields in the list. If this value is False, the field is not in the list and you cannot activating the track. There is also no possibility to change this value in the UI.
image

I do not see why this field needs to be persistent, but I agree changing it could also have unwanted side effects. But I did not have time yet to write the necessary upgrade script, to recalculate all values with the update. Without an update of the stored values, removing the readonly filter has no effect.

**Comment by CRogos** [comment]
> I m not sure about this. In a way, the behavior of readonly field is known and not changed by users. In order to avoid some noise with too many tracked fields by default, I will prefer to not include the readonly fields as tracked. IMHO, maybe just set as tracked the relevant fields to follow, as you mention in your specific case, will be enough.

I only give the user the option to track fields, and do not activate the tracking. (trackable vs. tracking_custom field)
Do you still have concerns?

**Comment by Kev-Roche** [comment]
> > I m not sure about this. In a way, the behavior of readonly field is known and not changed by users. In order to avoid some noise with too many tracked fields by default, I will prefer to not include the readonly fields as tracked. IMHO, maybe just set as tracked the relevant fields to follow, as you mention in your specific case, will be enough.
> 
> I only give the user the option to track fields, and do not activate the tracking. (trackable vs. tracking_custom field) Do you still have concerns?

My Bad, you are right, sorry I confound the two !

**Comment by Kev-Roche** [comment]
> > I m not sure about this. In a way, the behavior of readonly field is known and not changed by users. In order to avoid some noise with too many tracked fields by default, I will prefer to not include the readonly fields as tracked. IMHO, maybe just set as tracked the relevant fields to follow, as you mention in your specific case, will be enough.
> 
> I only give the user the option to track fields, and do not activate the tracking. (trackable vs. tracking_custom field) Do you still have concerns?

But if you set tracking on a new model or just update it with update_custom_tracking button, it will add all the readonly fields as tracked and we don't want too many automatic and unwanted tracked fields. Maybe just `trackable` as readonly=False and store will be enough to deal with specific cases, or making optional the readonly fields as trackable ?

**Comment by CRogos** [comment]
I've tested it on res.partner and there are 74 fields tracked instead of 59. Yes these are 15 fields more.
How about initializing the default domain this way to prevent readonly fields from automatic configuration?
image

**Comment by CRogos** [comment]
I've now re-added the store flag and added an upgrade script to recompute the existing value.

File: tracking_manager/models/ir_model.py [hunk 2]
```diff
@@ -73,8 +73,7 @@ def _get_model_tracked_by_o2m(self):
                 tracked_fields = custom_tracked_fields[model.model]
             else:
                 tracked_fields = model.field_id.filtered(
-                    lambda s: not s.readonly
-                    and not s.related
+                    lambda s: not s.related
```
Comment (Kev-Roche) [inline tracking_manager/models/ir_model.py]: It will activate the tracking on all readonly fields which are related to the main model by a o2m field. 
No need code here, for exemple a sale order with lines (o2M field). You have to activate the custom tracking on the sale.order.line model and select the pertinent fields to track.

File: tracking_manager/models/ir_model.py [hunk 3]
```diff
@@ -100,21 +99,27 @@ def _compute_automatic_custom_tracking(self):
...
             "account.move": [
+                ("readonly", "=", False),
                 "|",
                 ("ttype", "!=", "one2many"),
                 ("name", "in", ["invoice_line_ids"]),
             ],
-            "default_automatic_rule": [("ttype", "!=", "one2many")],
+            "default_automatic_rule": [
+                ("ttype", "!=", "one2many"),
+                ("readonly", "=", False),
```
Comment (Kev-Roche) [inline tracking_manager/models/ir_model.py:122]:
```suggestion
                ("readonly", "=", True),
```
or delete it, in order to prevent readonly fields from automatic tracking.

**Review by Kev-Roche — CHANGES_REQUESTED** [review]
as you said "to prevent readonly fields from automatic configuration", my main point is to keep the configuration and reading of tracking fields simple.
From my opinion, the tracking of readonly fields is marginal and can not be the main rule of this module. 

cc @sebastienbeau

File: tracking_manager/models/ir_model.py (hunk 3 above)
Comment (CRogos) [reply tracking_manager/models/ir_model.py:122]: Are you sure? This should prevent adding readonly fields by default.
image
image

File: tracking_manager/models/ir_model.py (hunk 2 above)
Comment (CRogos) [reply tracking_manager/models/ir_model.py]: I'll revert this change. I am not that deep into this part.

File: tracking_manager/models/ir_model.py (hunk 3 above)
Comment (Kev-Roche) [reply tracking_manager/models/ir_model.py:122]: My bad, you are right !

**Comment by CRogos** [comment]
@Kev-Roche I think everything is done now. Could you update your review to approved?

**Comment by CRogos** [comment]
I think everything is final now. Could you review again?

**Comment by CRogos** [comment]
@Kev-Roche @AnizR @legalsylvain could update your reviews?

**Comment by CRogos** [comment]
@Kev-Roche 
> From my opinion, the tracking of readonly fields is marginal and can not be the main rule of this module.

Unfortunately the account_asset module of odoo enterprise is one, were the fields are readonly and the user changes the fields via a wizard. So technically they are readonly, but actually they are not and they are very important to track.

If you really see a risk, I could also add a global setting that needs to be enabled first before allowing to track readonly attributes. But "Keep the configuration simple" could also mean not to add this global setting ;-) . 

@pedrobaeza do you have an opinion on this?

**Comment by Kev-Roche** [comment]
> @Kev-Roche
> 
> > From my opinion, the tracking of readonly fields is marginal and can not be the main rule of this module.
> 
> Unfortunately the account_asset module of odoo enterprise is one, were the fields are readonly and the user changes the fields via a wizard. So technically they are readonly, but actually they are not and they are very important to track.
> 
> If you really see a risk, I could also add a global setting that needs to be enabled first before allowing to track readonly attributes. But "Keep the configuration simple" could also mean not to add this global setting ;-) .
> 
> @pedrobaeza do you have an opinion on this?

I m good with the current way

**Review by Kev-Roche — APPROVED** [review]

**Comment by pedrobaeza** [comment]
I think it's usual that some fields are readonly, but unlocked on certain conditions (like state and having `force_save="1"` parameter) or modified by code, so they should be tracked as well.

**Comment by CRogos** [comment]
@sebastienbeau could you merge?

**Comment by CRogos** [comment]
@Kev-Roche do you have an idea what's the issue with the merge? Error does not look related to this change? Shall I rebase or could you give it another try?

**Comment by Kev-Roche** [comment]
@CRogos I have no Idea, don't see any related reason... can you rebase please, then I ll try again.

**Comment by CRogos** [comment]
This themes to be the problem: https://github.com/OCA/server-tools/pull/2578

**Comment by CRogos** [comment]
@Kev-Roche the problem with the repo has been solved. Could you retry the merge?
